### PR TITLE
[MM-356] Add feature to subscribe to release and workflow events

### LINKED ISSUE #565: Support for all github webhook events
Hi!

Can you add support for every github webhook [event](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads), or at least for GitHub Actions and Releases events?

### LINKED ISSUE #722: Feature Request: Subscribe to GitHub Action workflow failures
Some information we can include in the resulting post:

- Name of workflow
- Link to failed workflow run, and/or step that failed
- Link to any associated PR, or branch this was run on
- Possibly show a log of the run output associated with the failure, but this could be noisy and unhelpful if the logs at the end of the step failure are not related to the actual issue happening

### PR DESCRIPTION
**Pull request by ayusht2810**
#### Summary
- Added feature to publish workflow failure events

#### Screenshot 
image

#### Ticket Link
Fixes #722 
Fixes #565

#### What to test?
- Update the webhook for the repo to receive notifications for workflow events
- Create a subscription for a repo with a subscription enabled for workflow events
- Create workflow in which a particular jobs fail
- Check for the event received in the subscribed channel

#### Checklist
- [x] Completed dev testing
- [x] `make test` Ran test cases and ensured they are passing
- [x] `make check-style` Ran style check and ensured both webapp and server pass the checks

### CODE REVIEW

File: server/plugin/template.go [hunk 1]
```diff
@@ -429,6 +445,11 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 {{- else }} unstarred
 {{- end }} by {{template "user" .GetSender}}
 It now has **{{.GetRepo.GetStargazersCount}}** stars.`))
+
+	template.Must(masterTemplate.New("newWorkflowJob").Funcs(funcMap).Parse(`
+{{template "repo" .GetRepo}} {{template "workflowJob" .GetWorkflowJob}} workflow got failed (triggered by {{template "user" .GetSender}})
```
Comment (hanzei) [inline server/plugin/template.go:450]: Can we include the workflow name here?

File: README.md [hunk 2]
```diff
@@ -150,7 +150,7 @@ When you’ve tested the plugin and confirmed it’s working, notify your team s
    /github subscriptions add mattermost/mattermost-server --features issues,pulls,issue_comments,label:"Help Wanted"
    ```
   - The following flags are supported:
-     - `--features`: comma-delimited list of one or more of: issues, pulls, pulls_merged, pulls_created, pushes, creates, deletes, issue_creations, issue_comments, pull_reviews, label:"labelname". Defaults to pulls,issues,creates,deletes.
+     - `--features`: comma-delimited list of one or more of: issues, pulls, pulls_merged, pulls_created, pushes, creates, deletes, issue_creations, issue_comments, pull_reviews, workflows, label:"labelname". Defaults to pulls,issues,creates,deletes.
```
Comment (hanzei) [inline README.md:153]: I wonder if we should rename the flag to `workflow-failures` to make it clear that only these get notifications. Or add an option to `workflows` that specifies if only failed or also successful runs should trigger a notification.

File: server/plugin/webhook.go [hunk 3]
```diff
@@ -1377,3 +1385,44 @@ func (p *Plugin) postStarEvent(event *github.StarEvent) {
...
+		if !sub.Workflows() {
+			continue
+		}
+
+		post := &model.Post{
+			UserId:    p.BotUserID,
+			Type:      "custom_git_workflow_job",
+			Message:   newWorkflowJobMessage,
+			ChannelId: sub.ChannelID,
+		}
```
Comment (hanzei) [inline server/plugin/webhook.go:1422]: Multiple subs should not share a post, see https://github.com/mattermost/mattermost-plugin-github/pull/755

**Review by hanzei — CHANGES_REQUESTED** [review]

File: server/plugin/webhook.go [hunk 4]
```diff
@@ -1377,3 +1385,44 @@ func (p *Plugin) postStarEvent(event *github.StarEvent) {
...
+	newWorkflowJobMessage, err := renderTemplate("newWorkflowJob", event)
+	if err != nil {
+		p.client.Log.Warn("Failed to render template", "Error", err.Error())
+		return
+	}
+
+	for _, sub := range subs {
+		if !sub.Workflows() {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/webhook.go:1415]: Makes sense to me :+1: 

Limiting the notifications to a specific branch seems like a common use case. Can we include that in there as well?

[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/plugin/webhook.go:1415]: > `Just the ci workflow`

@mickmister Are you suggesting here to add a subscription on the basis of the workflow name, which is `ci` in the above case?

[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/plugin/webhook.go:1415]: > Limiting the notifications to a specific branch seems like a common use case. Can we include that in there as well?

@mickmister @hanzei For the above case, what should be the slash command format? Should we keep something like: `/github subscriptions add owner/repo pulls,workflows:branch-b1,b2;name-n1,n2;job-j1,j2`? Please let me know if you have any better suggestions for the above case.

[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/webhook.go:1415]: The `subscriptions add` feature is quite overloaded at this point. Adding more complexity on top of it may not be the best thing to do. Created a ticket for breaking this out of the command https://github.com/mattermost/mattermost-plugin-github/issues/780

Mainly thinking of the complexity of the `job` parameter there. I'm thinking we should avoid the `job` parameter, since those jobs may not belong to all workflows. Then it's up to the actions workflow developer in that project to structure the workflows in a way to be something to subscribe to.

File: README.md (hunk 2 above)
Comment (mickmister) [reply README.md]: I'm thinking we should have separate `workflow-success` and `workflow-failure` subscription types

Comment (Kshitij-Katiyar) [reply README.md]: @mickmister Added the `workflow_success`. Please have a look

File: go.mod [hunk 5]
```diff
@@ -13,13 +13,16 @@ require (
 	golang.org/x/oauth2 v0.17.0
 )
 
+require (
+	github.com/ProtonMail/go-crypto v0.0.0-20230217124315-7d5c6f04bbb8 // indirect
+	github.com/cloudflare/circl v1.3.7 // indirect
+)
+
 require (
```
Comment (mickmister) [inline go.mod:21]: Can we put all this in the same `require` block?

File: server/plugin/command.go [hunk 6]
```diff
@@ -896,7 +900,7 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
 
 	subscriptionsAdd := model.NewAutocompleteData("add", "[owner/repo] [features] [flags]", "Subscribe the current channel to receive notifications about opened pull requests and issues for an organization or repository. [features] and [flags] are optional arguments")
 	subscriptionsAdd.AddTextArgument("Owner/repo to subscribe to", "[owner/repo]", "")
-	subscriptionsAdd.AddNamedTextArgument("features", "Comma-delimited list of one or more of: issues, pulls, pulls_merged, pulls_created, pushes, creates, deletes, issue_creations, issue_comments, pull_reviews, workflows, label:\"<labelname>\". Defaults to pulls,issues,creates,deletes", "", `/[^,-\s]+(,[^,-\s]+)*/`, false)
+	subscriptionsAdd.AddNamedTextArgument("features", "Comma-delimited list of one or more of: issues, pulls, pulls_merged, pulls_created, pushes, creates, deletes, issue_creations, issue_comments, pull_reviews, workflow_success, workflow_failures, label:\"<labelname>\". Defaults to pulls,issues,creates,deletes", "", `/[^,-\s]+(,[^,-\s]+)*/`, false)
```
Comment (mickmister) [inline server/plugin/command.go:903]: Let's go with "singular" for both feature types:

- `workflow_success`
- `workflow_failure`

Comment (mickmister) [reply server/plugin/command.go:903]: Do we need to mention `releases` as well?

File: webapp/package.json [hunk 7]
```diff
@@ -58,7 +58,7 @@
     "sass": "1.60.0",
     "sass-loader": "10.4.1",
     "style-loader": "1.2.1",
-    "typescript": "4.0.3",
+    "typescript": "4.2.3",
```
Comment (mickmister) [inline webapp/package.json:61]: Is this needed for this PR? As well as the changes to `package-lock.json`?

File: server/plugin/webhook.go [hunk 8]
```diff
@@ -1382,3 +1388,47 @@ func (p *Plugin) postWorkflowJobEvent(event *github.WorkflowJobEvent) {
...
+	}
+
+	repo := event.GetRepo()
+	subs := p.GetSubscribedChannelsForRepository(repo)
+
+	if len(subs) == 0 {
+		return
+	}
+
+	newReleaseMessage, err := renderTemplate("newReleaseEvent", event)
```
Comment (mickmister) [inline server/plugin/webhook.go:1412]: @phoinixgrr FYI this feature in the GitHub plugin will allow us to create subscriptions to new releases

File: server/plugin/template.go [hunk 9]
```diff
@@ -447,9 +454,15 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 It now has **{{.GetRepo.GetStargazersCount}}** stars.`))
 
 	template.Must(masterTemplate.New("newWorkflowJob").Funcs(funcMap).Parse(`
-{{template "repo" .GetRepo}} {{template "workflowJob" .GetWorkflowJob}} workflow got failed (triggered by {{template "user" .GetSender}})
-Step failed: {{.GetWorkflowJob.Steps | workflowJobFailedStep}}
+{{template "repo" .GetRepo}} {{.GetWorkflowJob.GetWorkflowName}} workflow {{if eq .GetWorkflowJob.GetConclusion "success"}}succeeded{{else}}failed{{end}} (triggered by {{template "user" .GetSender}})
+{{if eq .GetWorkflowJob.GetConclusion "failure"}}Job failed: {{template "workflowJob" .GetWorkflowJob}}
+Step failed: {{.GetWorkflowJob.Steps | workflowJobFailedStep}}{{end}}
```
Comment (mickmister) [inline server/plugin/template.go:459]: Can you explain how this works? How do we know which job failed?

Also what is going on with `{{.GetWorkflowJob.Steps | workflowJobFailedStep}}`? I'm not sure what `workflowJobFailedStep` is

**Review by mickmister — COMMENTED** [review]
@Kshitij-Katiyar Great job on this :+1: I have a few more comments

File: webapp/package.json (hunk 7 above)
Comment (Kshitij-Katiyar) [reply webapp/package.json:61]: @mickmister All the changes in package-lock.json are because of the upgrade in typescript version. The upgrade is necessary to pass the `ci-lint` check.

File: server/plugin/command.go (hunk 6 above)
Comment (Kshitij-Katiyar) [reply server/plugin/command.go:903]: @mickmister I think we should

File: server/plugin/template.go (hunk 9 above)
Comment (Kshitij-Katiyar) [reply server/plugin/template.go:459]: @mickmister 
The line `{{if eq .GetWorkflowJob.GetConclusion "failure"}}` checks if the workflow job’s conclusion is `"failure"`. If the job has indeed failed, the template renders additional details about the job using the workflowJob template: `Job failed: {{template "workflowJob" .GetWorkflowJob}}`.

The `workflowJobFailedStep` function processes the Steps array from the `WorkflowJob` object. It filters these steps to find and return details about the specific step that failed. This function is defined in the `funcMap` provided to the template engine, allowing it to be used as a filter within the template.

**Review by mickmister — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (wiggin77) [reply README.md]: @hanzei are you still requesting changes here?

Comment (hanzei) [reply README.md]: Where did the `README.md` updates go?

File: server/plugin/template_test.go [hunk 10]
```diff
@@ -1477,6 +1477,45 @@ func TestGitHubUsernameRegex(t *testing.T) {
 	}
 }
 
+func TestWorkflowJobNotification(t *testing.T) {
```
Comment (hanzei) [inline server/plugin/template_test.go:1480]: Can you please also add a test case of the successful workflow?

File: webapp/package-lock.json
Comment (hanzei) [inline webapp/package-lock.json:1]: Unrelated change?

**Comment by wiggin77** [comment]
This PR is going to have a significant merge conflict with https://github.com/mattermost/mattermost-plugin-github/pull/808/files 

Does anyone have an opinion on which is more important?: release and workflow subscriptions vs discussions and comment

**Review by hanzei — APPROVED** [review]

**Comment by raghavaggarwal2308** [comment]
@wiggin77 In both the PRs we mainly adding new functions not updating the existing functions much. I think it won't be difficult to fix the conflicts in this case. So, we can merge whichever PR is ready first. Please let me know your opinions on this.

**Comment by wiggin77** [comment]
> @wiggin77 In both the PRs we mainly adding new functions not updating the existing functions much. I think it won't be difficult to fix the conflicts in this case. So, we can merge whichever PR is ready first. Please let me know your opinions on this.

Ok, let's get this one merged first.

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:

**Review by arush-vashishtha — APPROVED** [review]
Tested this PR for both success and failure scenarios in the workflow and the user is receiving the message. LGTM.
